lib/elixir/lib/uri.ex

```diff
@@ -373,6 +373,65 @@ defmodule URI do
       "http://google.com"
   """
   defdelegate to_string(uri), to: String.Chars.URI
+
+  @doc ~S"""
+  Merges two URIs as per RFC 3986 §5.2
+
+  ## Examples
+
+      iex> URI.merge(URI.parse("http://google.com"), "/query") |> to_string
+      "http://google.com/query"
+
+      iex> URI.merge("http://example.com", "http://google.com")|> to_string
+      "http://google.com"
+  """
+  def merge(%URI{authority: nil}, %URI{}), do: raise ArgumentError, "You must merge onto an absolute URI"
+  def merge(%URI{}, %URI{scheme: scheme}=rel) when not is_nil(scheme) do
+    rel
+  end
+  def merge(%URI{}=base, %URI{path: "", query: query, fragment: fragment}=rel) do
+    merge(base, %{rel | path: nil, query: query, fragment: fragment})
+  end
+  def merge(%URI{}=base, %URI{path: nil, query: query, fragment: fragment}) do
+    %{base | path: base.path, query: query || base.query, fragment: fragment}
+  end
+  def merge(%URI{path: base_path}=base, %URI{path: path, query: query, fragment: fragment}) do
+    %{base | path: merge_paths(base_path, path), query: query, fragment: fragment}
+  end
+  # Convert binary to URI
+  def merge(base, rel), do: merge(URI.parse(base), URI.parse(rel))
+
+  defp merge_paths(nil, rel_path), do: merge_paths("/", rel_path)
+  defp merge_paths(_, << "/" <> _ >>=rel_path), do: rel_path
+  defp merge_paths(base_path, rel_path) do
+    [_ | base_segments] = path_to_segments(base_path)
+    rel_segments = path_to_segments(rel_path)
+    rel_segments ++ base_segments
+    |> remove_dot_segments
+    |> Enum.join("/")
+  end
+
+  defp remove_dot_segments(list, acc \\ [], repeat \\ false)
+  defp remove_dot_segments([], acc, true), do: remove_dot_segments(Enum.reverse(acc))
+  defp remove_dot_segments([], ["" | _]=acc, false), do: acc
+  defp remove_dot_segments([], acc, false), do: ["" | acc]
+  defp remove_dot_segments(["." | tail], acc, repeat) do
+    remove_dot_segments(tail, acc, repeat)
+  end
+  defp remove_dot_segments([".." | [".." | _]=tail], acc, _repeat) do
```

Comment:
Instead of using repeat, I believe you could implement this by taking items from the accumulator. Here is a more concise implementation of removing dots: https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/path.ex#L644-L658